File: HW07/Rectangle.java
import java.awt.Color;
import java.awt.Graphics;
import java.lang.IllegalArgumentException;

/**
 * 
 * A simple Rectangle class that 
 * implements shape methods.
 * 
 * @author Yusuf Can Kan
 *
 */



public class Rectangle implements Shape,Cloneable {

	/**
	 * Width for rectangle.
	 */
	private double width;
	
	/**
	 * Height for rectangle.
	 */
	private double height;
	
	/**
	 * X coordinate.
	 */
	private double x;
	
	/**
	 * Y coordinates.
	 */
	private double y;
	
	/**
	 * The area of rectangles which created so far.
	 */
	private static double total_area=0;
	
	/**
	 * The perimeter of rectangles which created so far.
	 */
	private static double total_perimeter=0;
	
	/**
	 * Simple constructor that sets width,height,x and y coordinates.
	 * @param width_ width of rectangle.
	 * @param height_ height of rectangle.
	 * @param x_ x coordinate of rectangle.
	 * @param y_ y coordinate of rectangle.
	 * @throws IllegalArgumentException throws exception if width and height is smaller than 0.
	 */
	public Rectangle(double width_,double height_,double x_,double y_)
			throws IllegalArgumentException{
		
		if(width_< 0 || height<0 )
			throw new IllegalArgumentException();
		
		width=width_;
		height=height_;
		x=x_;
		y=y_;
		this.increase_total_area();
		this.increase_total_perimeter();
	}
	
	/**
	 * Simple constructor that sets width and height.
	 * @param width_ width of rectangle.
	 * @param height_ height of rectangle.
	 */
	public Rectangle(double width_,double height_){
		this(width_,height_,0,0);
	}
	
	/**
	 * Simple constructor that sets width and height the same value.
	 * @param width_ width and height of the rectangle.
	 */
	public Rectangle(double width_){
		this(width_,width_,0,0);
	}
	
	/**
	 * Simple constructor that sets rectangle lengths 20.
	 */
	public Rectangle(){
		this(20,20,0,0);
	}
	
	/**
	 * Getter method for width.
	 * @return width of rectangle.
	 */
	public double getWidth(){return width;}
	
	/**
	 * Getter method for height.
	 * @return height of rectangle.
	 */
	public double getHeight(){return height;}
	
	/**
	 * Getter method for x coordinates.
	 * @return x coordinates.
	 */
	public double getX() {return x;}
	
	/**
	 * Getter method for y coordinates.
	 * @return y coordinates.
	 */
	public double getY(){return y;}
	
	/**
	 * Setter method for width of rectangle.
	 * @param width_ width of rectangle
	 * @throws IllegalArgumentException throws if width is smaller than 0.
	 */
	public void setWidth(double width_)
			throws IllegalArgumentException{
		if(width_ < 0){
			throw new IllegalArgumentException();
		}
		
		width = width_;
	}
	
	/**
	 * Setter method for height of rectangle.
	 * @param height_ height of rectangle.
	 * @throws IllegalArgumentException throws if height is smaller than 0.
	 */
	public void setHeight(double height_)
			throws IllegalArgumentException{
		if(height_ < 0){
			throw new IllegalArgumentException();
		}
		
		height = height_;
	}
	
	/**
	 * Setter method for x coordinate.
	 * @param x_ x coordinate.
	 * @throws IllegalArgumentException throws if x < 0.
	 */
	public void setX(double x_)
			throws IllegalArgumentException{
		if(x_<0)
			throw new IllegalArgumentException();
		
		x = x_;
	}
	
	/**
	 * Setter method for y coordinates.
	 * @param y_ y coordinate.
	 * @throws IllegalArgumentException throws if y<0.
	 */
	public void setY(double y_)
			throws IllegalArgumentException{
		if(y_<0)
			throw new IllegalArgumentException();
		
		y = y_;
	}
	
	@Override
	public double area() {
		return width*height;
	}

	@Override
	public double perimeter() {
		return 2*width+2*height;
	}

	@Override
	public void increment() {
		x++;
		y++;
	}

	@Override
	public void decrement() {
		if( (x-1) >= 0) x--;
		if( (y-1) >= 0) y--;
	}

	@Override
	public void draw(Graphics g) {
		g.setColor(Color.red);
		g.fillRect((int)x, (int)y, (int)width, (int)height);
		g.setColor(Color.black);
		g.drawRect((int)x, (int)y, (int)width, (int)height);
	}

	@Override
	public int compareTo(Shape o) {
		if( o.area() == this.area() ) return 1;
		else return 0;
	}
	
	/**
	 * Increases total area.
	 */
	public void increase_total_area(){
		total_area+= width*height;
	}
	
	/**
	 * Increases total perimeter.
	 */
	public void increase_total_perimeter(){
		total_perimeter += 2*width+2*height;
	}
	
	/**
	 * returns the total area.
	 * @return total area.
	 */
	public double show_total_area(){
		return total_area;
	}
	
	/**
	 * Returns the total perimeter.
	 * @return total perimeter.
	 */
	public double show_total_perimeter(){
		return total_perimeter;
	}
	
	/**
	 * Clone method.
	 */
	public Rectangle clone() 
			throws CloneNotSupportedException{ 
		return (Rectangle)super.clone(); 
	}
	
	

}

File: HW05/Rectangle.cpp
#include "Rectangle.h"

namespace MyShape{

	
	double Rectangle::total_area=0;
	double Rectangle::total_perimeter=0;

	Rectangle::Rectangle(double Width,double Height):Shape(),width(Width),height(Height){ }
	Rectangle::Rectangle(double Width):Shape(),width(Width),height(1.){  }

	Rectangle::Rectangle():Shape(),width(1.),height(1.){ }

	double Rectangle::area()const{ return height*width; }
	double Rectangle::perimeter()const{ return 2*height+2*width; }


	double Rectangle::getWidth()const throw(){ return width; }
	double Rectangle::getHeight()const throw(){ return height; }
	double Rectangle::getX() const throw(){ return x; }
	double Rectangle::getY() const throw(){ return y; }

	void Rectangle::setWidth(double Width) throw(ExceptionErrorrr){ 
		try{
			if(Width < 0)
				throw ExceptionErrorrr(Width);

			width=Width; 
		}
		catch(ExceptionErrorrr a){
			cout<<endl<<"width:"<<a.getMessage()<<" cannot be negative";
		}
	}
	void Rectangle::setHeight(double Height) throw(ExceptionErrorrr){ 
		try{
			if(Height <0)
				throw ExceptionErrorrr(Height);
		}
		catch(ExceptionErrorrr a){
			cout<<endl<<"Height:"<<a.getMessage()<<" cannot be negative";
		}
		height=Height;

	}
	void Rectangle::setX(double _x) throw(ExceptionErrorrr){ 
		try{
			if(_x<0){	
				throw ExceptionErrorrr(_x);
			}
			x=_x; 
		}
		catch(ExceptionErrorrr a){
			cout<<endl<<"x:"<<a.getMessage()<<" cannot be negative";
		}
	}
	void Rectangle::setY(double _y) throw(ExceptionErrorrr){ 
		try{
			if(_y<0)
				throw ExceptionErrorrr(_y);
		 y=_y; 
		}
		catch(ExceptionErrorrr a){
			cout<<endl<<"y:"<<a.getMessage()<<" cannot be negative";
		}
	}

	Shape& Rectangle::operator++() throw(){
		x++;
		y++;
		return (*this);
	}

	Shape& Rectangle::operator++(int IgnoreMe) throw(){
		Rectangle *rectangle = new Rectangle(*this);
		x++;
		y++;
		return (*rectangle);
	}

	Shape& Rectangle::operator--() throw(){	
				x--;
				y--;
			return (*this);
	}

	Shape& Rectangle::operator--(int IgnoreMe) throw(){		
		Rectangle *rectangle = new Rectangle(*this);	
				x--;
				y--;
			return (*rectangle);
	}

	ostream& operator<<(ostream &outputStream,Rectangle rectangle){
		outputStream << "<rect x=\""<< rectangle.getX() << "\" y=\""<< rectangle.getY() <<"\" width=\""<< rectangle.getWidth() <<"\" height=\""
							<< rectangle.getHeight() <<"\" fill=\"blue\" style=\"fill:blue;stroke:pink;stroke-width:5\"/>" << endl;
		return outputStream;
	}

	void Rectangle::draw(ostream &outputStream) throw(){
		outputStream << "<rect x=\""<< x << "\" y=\""<< y <<"\" width=\""<< width <<"\" height=\""
					<< height <<"\" fill=\"blue\" style=\"fill:blue;stroke:pink;stroke-width:5\"/>" << endl;
	}

	/*for static total_area and total_perimeter.*/
	void Rectangle::increase_total_area(){
		total_area+=width*height;
	}
	void Rectangle::increase_perimeter(){
		total_perimeter=(2*width)+(2*height);
	}
	double Rectangle::show_total_area(){
		return total_perimeter;
	}
	double Rectangle::show_total_perimeter(){
		return total_area;
	}

}





